File: a-injection-SQL/a-injection-SQL-corrige.php
<?php

if(isset($_POST['signin'])){
    $pdo = new PDO("mysql:host=localhost;dbname=securite", 'root', '');
    // AJOUT de la fonction addslashes devant la variable login recupérée du formulaire
    $login = addslashes($_POST['login']);
    $pwd = $_POST['pwd'];
    $selectall = $pdo->query("SELECT * FROM user WHERE login='$login' AND pwd='$pwd'");
    $result = $selectall->fetch();
    $counttable = count((is_countable($result)?$result:[]));
    if($counttable!=0){
        echo 'connexion réussie';
    }
    else{
        echo 'utilisateur non reconnu';
    }
}

?>

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <h1>Injection SQL</h1>

    <form method="post" action="#">
        Login : <input type="text" name="login"><br>
        Pwd : <input type="text" name="pwd"><br>
        <input type="submit" value="Sign in" name="signin">
    </form>
    <p>Essayer avec le login <strong>' OR 1=1 OR 1='</strong> : la connexion ne fonctionne plus :-)</p>
</body>
</html>

File: a-injection-SQL/a-injection-SQL-corrige.py
import MySQLdb
db=MySQLdb.connect(host="localhost",user="root",password="",database="securite")
c=db.cursor()
c.execute("SELECT * FROM user WHERE login='admin'")
result = c.fetchone()

def is_admin(username: str) -> bool:
    with db.cursor() as cursor:
        cursor.execute("""
            SELECT
                admin
            FROM
                user
            WHERE
                login = %(username)s
        """, {
            'username': username
        })
        result = cursor.fetchone()

    if result is None:
        # User does not exist
        return False

    admin, = result
    return admin

print(is_admin('admin'))

print(is_admin('test'))
is_admin("'; update user set admin = 'true' where login = 'test';")

# Le code malveillant ne fonctionne plus

# https://realpython.com/prevent-python-sql-injection/